### PR TITLE
Qt RPC console: history sensitive-data filter, and saving input line when browsing history

### PR DESCRIPTION
**Pull request by luke-jr**
Rescuing uncontroversial parts of #5891

Alternative to #8746

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Comment by sipa** [comment]
Concept ACK. Maybe add a comment that signrawtransaction can be removed from this list if it's ever split into a wallet call and a utility call.

**Comment by achow101** [comment]
If we're adding privkey stuff here, then `signmessagewithprivkey` needs to be added.

**Comment by paveljanik** [comment]
Concept ACK

**Comment by luke-jr** [comment]
@sipa If `signrawtransaction` is split, there will likely still be users trying to use the old usage for at least one release, so I'm not sure it makes sense to remove it from the filter, at least not at the same time?

@achow101 I agree it would be better to add dummy history items, but comments on previous PRs seem to suggest that is a source of disagreement.

Added `signmessagewithprivkey` to the filter.

**Comment by achow101** [comment]
@luke-jr instead of masking out the arguments as I did in my PR, what if you just added the command name to the history. Since all the commands pass through that filter method, you could have it return the string that goes into the history. For non-filtered commands, it returns the command itself. For filtered ones, it just returns the command name.

**Comment by laanwj** [comment]
> Instead of masking out the arguments as I did in my PR, what if you just added the command name to the history. 

Would work for me. My main comment on doing it @achow101 's way with masking individual arguments is just that that is too brittle and hard to maintain. If having just the command name to the history is useful in any way, which I doubt a bit with the incredible autocompletion that the debug console has these days, you could do that.

**Comment by luke-jr** [comment]
I consider it useful to have *any* placeholder in history, since otherwise one might <up>-<enter> and get the second-to-last execution instead. At least with a dummy placeholder, they get an error/help.

**Comment by jonasschnelli** [comment]
1. The passphrase is still visible in the console as executing command when calling `walletpassphrase` (should this be done in a different pull request?)
2. Nested commands makes this a little bit more complex. Executing "a dump" `getnewaddress(walletpassphrase(test))` results in executing `walletpassphrase(test)` but actually not "hiding" the passphrase from the history.


bildschirmfoto 2016-10-09 um 10 15 55

**Comment by luke-jr** [comment]
Rebased and taught it to handle nested commands. When a filtered command is encountered, all its parameters are replaced with "(…)" in the command history. Unit tests now check this.

File: src/qt/rpcconsole.cpp [hunk 1]
```diff
@@ -63,6 +64,19 @@ const struct {
...
+
+// don't add private key handling cmd's to the history
+const QStringList historyFilter = QStringList()
+    << "importprivkey"
+    << "signmessagewithprivkey"
+    << "signrawtransaction"
+    << "walletpassphrase"
+    << "walletpassphrasechange"
+    << "encryptwallet";
+
```
Comment (jonasschnelli) [inline src/qt/rpcconsole.cpp]: nit: newline

**Comment by jonasschnelli** [comment]
Tested a bit.
Needs coverage for `importmulti` now.

File: src/qt/rpcconsole.cpp (hunk 1 above)
Comment (luke-jr) [inline src/qt/rpcconsole.cpp]: ?

**Comment by jonasschnelli** [comment]
Looks good. Squash/combine some commits?

**Comment by luke-jr** [comment]
Rather not squash... They look reasonably logical progression, and squashing is a bad practice anyway.

**Comment by jonasschnelli** [comment]
Tested ACK 8562792095a78fdef2dacc3c5e32c41206df4c87

**Comment by btcdrak** [comment]
Tested ACK 8562792095a78fdef2dacc3c5e32c41206df4c87

**Comment by paveljanik** [comment]
`Wshadow` statistics: 
   1 qt/rpcconsole.cpp:173:56: warning: declaration shadows a local variable [-Wshadow]

**Comment by paveljanik** [comment]
Why do you think so? It is clear that `curarg` inside the block shadows `curarg` outside.

**Comment by jonasschnelli** [comment]
@paveljanik: Maybe write a fix PR?

**Comment by luke-jr** [comment]
@paveljanik It's a different scope, and lambda functions do not inherit the scope of the calling function beyond what you tell it to.

**Comment by sipa** [comment]
@luke-jr Not by default, but your lambda has capture list `[&]`, which means it does capture everything.

**Comment by sipa** [comment]
@luke-jr Well, yes. But if -Wshadow wants to catch potential errors resulting from variables being identically named, it should absolutely catch this.

**Comment by paveljanik** [comment]
@luke-jr What name do you want me to use inside lambda instead of `curarg`? `strArg`?

**Comment by luke-jr** [comment]
Sure

**Pull request merged**